lib/fog/xenserver/compute/real.rb

```diff
@@ -5,22 +5,24 @@ class Real
         attr_reader :host, :username
 
         def initialize(options={})
-          @host        = options[:xenserver_url]
-          @username    = options[:xenserver_username]
-          @password    = options[:xenserver_password]
-          @defaults    = options[:xenserver_defaults] || {}
-          @timeout     = options[:xenserver_timeout] || 30
-          @use_ssl     = options[:xenserver_use_ssl] || false
-          @port        = options[:xenserver_port] || 80
-          @verify_mode = options[:xenserver_verify_mode] || OpenSSL::SSL::VERIFY_PEER
+          @host               = options[:xenserver_url]
+          @username           = options[:xenserver_username]
+          @password           = options[:xenserver_password]
+          @defaults           = options[:xenserver_defaults] || {}
+          @timeout            = options[:xenserver_timeout] || 30
+          @use_ssl            = options[:xenserver_use_ssl] || false
+          @port               = options[:xenserver_port] || 80
+          @verify_mode        = options[:xenserver_verify_mode] || OpenSSL::SSL::VERIFY_PEER
           @connection  = Fog::XenServer::Connection.new(
             @host, @port, @use_ssl, @verify_mode, @timeout)
 
-          @connection.authenticate(@username, @password)
-
-          if @connection.slave?
-            @connection = Fog::XenServer::Connection.new(@connection.master, @port, @use_ssl, @verify_mode, @timeout)
+          begin
             @connection.authenticate(@username, @password)
+          rescue Fog::XenServer::HostIsSlave => e
+              @connection = Fog::XenServer::Connection.new(
+                e.master, @port, @use_ssl, @verify_mode, @timeout
+              )
+              @connection.authenticate(@username, @password)
```

Comment:
Maybe we should push this logic to the `Connection` class?

Reply:
Could you explain more your idea ? Do you think I have to renew the XMLRPC client instance in the authenticate method of the Connection class ?

Reply:
Yeap. Thinking about it i believe this code is related to the act of connect to XenServer and since we are talking about reconnecting to master automatically in case of a connection to a slave, i believe it make sense to put it together.

This way the connection (and reconnection in case it needs) will be completely transparent to the final user and to `Compute` service itself.

WDYT?

Reply:
I have 2 suggestions :

1. In the initialize method of the Connection class after the initialization of the factory, we try to login (session.login_with_password) without credentials and we reconnect if the error is HOST_IS_SLAVE. Like that : 
` 
  def initialize(host, port, use_ssl, verify_mode, timeout)
        @factory = connect(host, port, use_ssl, verify_mode, timeout)
        response = @factory.call("session.login_with_password", "", "")
        unless response["Status"] =~ /Success/
          if response.key?("ErrorDescription") &&
             response["ErrorDescription"].is_a?(Array) &&
             response["ErrorDescription"].length >= 2 &&
             response["ErrorDescription"][0] == "HOST_IS_SLAVE"
            host = response["ErrorDescription"][1]
            @factory = connect(host, port, use_ssl, verify_mode, timeout)
          end
        end
      end

      def connect(host, port, use_ssl, verify_mode, timeout)
        factory = XMLRPC::Client.new3(host: host, port: port, use_ssl: use_ssl, path: "/")
        if factory.respond_to?(:http)
          factory.http.verify_mode = verify_mode
        else
          factory.instance_variable_get(:@http).verify_mode = verify_mode
        end
        factory.set_parser(NokogiriStreamParser.new)
        factory.timeout = timeout
        factory
      end

     def authenticate( username, password )
        response = @factory.call("session.login_with_password", username.to_s, password.to_s)
        unless response["Status"] =~ /Success/
          raise Fog::XenServer::InvalidLogin.new
        end
        @credentials = response["Value"]
      end
`

2. The second solution is to have the reconnection in the authenticate method. It's less logic but it needs an API call of less. Like that : 
`
      def initialize(host, port, use_ssl, verify_mode, timeout)
        @host = host
        @port = port
        @use_ssl = use_ssl
        @verify_mode = verify_mode
        @timeout = timeout
        connect()
      end

      def connect()
        @factory = XMLRPC::Client.new3(host: @host, port: @port, use_ssl: @use_ssl, path: "/")
        if @factory.respond_to?(:http)
          @factory.http.verify_mode = @verify_mode
        else
          @factory.instance_variable_get(:@http).verify_mode = @verify_mode
        end
        @factory.set_parser(NokogiriStreamParser.new)
        @factory.timeout = @timeout
      end

      def authenticate( username, password )
        response = @factory.call("session.login_with_password", username.to_s, password.to_s)
        unless response["Status"] =~ /Success/
          if response.key?("ErrorDescription") &&
             response["ErrorDescription"].is_a?(Array) &&
             response["ErrorDescription"].length >= 2 &&
             response["ErrorDescription"][0] == "HOST_IS_SLAVE"
            @host = response["ErrorDescription"][1]
            connect()
            response = @factory.call("session.login_with_password", username.to_s, password.to_s)
            unless response["Status"] =~ /Success/
              raise Fog::XenServer::InvalidLogin.new
            end
          end
          raise Fog::XenServer::InvalidLogin.new
        end
        @credentials = response["Value"]
      end
`

WDYT ?

Reply:
I believe it makes more sense in the `authenticate` method because this is gonna be an implied logic to authenticate IMO. So i believe second option. Agree?

Reply:
Ok so what do you think of my latest commit ?